processor/transformprocessor/internal/traces/func_set_semconv_span_name.go

```diff
@@ -0,0 +1,256 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package traces // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/transformprocessor/internal/traces"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.opentelemetry.io/otel/attribute"
+	semconv "go.opentelemetry.io/otel/semconv/v1.37.0"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/ottlspan"
+)
+
+type setSemconvSpanNameArguments struct {
+	SemconvVersion            string
+	OriginalSpanNameAttribute ottl.Optional[string]
+}
+
+func NewSetSemconvSpanNameFactory() ottl.Factory[ottlspan.TransformContext] {
+	return ottl.NewFactory("set_semconv_span_name", &setSemconvSpanNameArguments{}, createSetSemconvSpanNameFunction)
+}
+
+func createSetSemconvSpanNameFunction(_ ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[ottlspan.TransformContext], error) {
+	args, ok := oArgs.(*setSemconvSpanNameArguments)
+
+	if !ok {
+		return nil, errors.New("NewSetSemconvSpanNameFactory args must be of type *setSemconvSpanNameArguments")
+	}
+	semconvVersion := args.SemconvVersion
+	originalSpanNameAttribute := args.OriginalSpanNameAttribute
+
+	// Currently only v1.37.0 is supported
+	supportedSemconvVersion := "1.37.0"
+	if semconvVersion != supportedSemconvVersion {
+		return nil, fmt.Errorf("unsupported semconv version: %s, supported version: %s", semconvVersion, supportedSemconvVersion)
+	}
+
+	return func(_ context.Context, tCtx ottlspan.TransformContext) (any, error) {
+		setSemconvSpanName(originalSpanNameAttribute, tCtx.GetSpan())
+		return nil, nil
+	}, nil
+}
+
+func setSemconvSpanName(originalSpanNameAttribute ottl.Optional[string], span ptrace.Span) {
+	originalSpanName := span.Name()
+	semConvSpanName := SemconvSpanName(span)
+	span.SetName(semConvSpanName)
+	if originalSpanName != semConvSpanName && originalSpanNameAttribute.GetOr("") != "" {
```

Comment:
We should probably validate it at function's creation time to not allow `originalSpanNameAttribute` to be set to empty string (`!originalSpanNameAttribute.IsEmpty() && originalSpanNameAttribute.Get() == ""`).  Please note that the `IsEmpty()` function returns `true` if the `Optional[T]` does not contain a value, and not if it's an empty string.